frontends/parsers/p4/p4lexer.ll

```diff
@@ -0,0 +1,199 @@
+%{
+#include "frontends/common/constantParsing.h"
+#include "frontends/parsers/parserDriver.h"
+#include "frontends/parsers/p4/p4lexer.hpp"
+#include "frontends/parsers/p4/p4parser.hpp"
+
+using Parser = P4::P4Parser;
+
+#undef  YY_DECL
+#define YY_DECL Parser::symbol_type P4::P4Lexer::yylex(P4::P4ParserDriver& driver)
+
+#define YY_USER_ACTION driver.onReadToken(yytext);
+#define yyterminate() return Parser::make_END(driver.yylloc);
+
+// Silence the warnings triggered by the code flex generates.
+#pragma GCC diagnostic ignored "-Wpragmas"
+#pragma GCC diagnostic ignored "-Wunused-function"
+#pragma GCC diagnostic ignored "-Wunused-variable"
+#pragma GCC diagnostic ignored "-Wsign-compare"
+#pragma GCC diagnostic ignored "-Wtautological-undefined-compare"
+
+%}
+
+%option c++
+%option outfile="lex.yy.c"
+%option yyclass="P4::P4Lexer"
+%option prefix="p4"
+%option nodefault noyywrap nounput noinput noyyget_leng
+%option noyyget_debug noyyset_debug noyyget_extra noyyset_extra noyyget_in noyyset_in
+%option noyyget_out noyyset_out noyyget_text noyyget_lineno noyyset_lineno
+
+%x COMMENT STRING
+%x LINE1 LINE2 LINE3
+%s NORMAL
+
+%%
+
+[ \t\r]+          ;
+[\n]            { BEGIN INITIAL; }
+"//".*            ;
+"/*"            { BEGIN COMMENT; }
+<COMMENT>"*/"   { BEGIN NORMAL; }
+<COMMENT>.        ;
+<COMMENT>[\n]     ;
+
+<INITIAL>"#line"      { BEGIN(LINE1); }
+<INITIAL>"# "         { BEGIN(LINE1); }
+<INITIAL>[ \t]*"#"    { BEGIN(LINE3); }
+<LINE1>[0-9]+         { BEGIN(LINE2); driver.onReadLineNumber(yytext); }
+<LINE2>\"[^\"]*       { BEGIN(LINE3); driver.onReadFileName(yytext+1); }
+<LINE1,LINE2>[ \t]      ;
+<LINE1,LINE2>.        { BEGIN(LINE3); }
+<LINE3>.                ;
+<LINE1,LINE2,LINE3>\n { BEGIN(INITIAL); }
+<LINE1,LINE2,LINE3,COMMENT,NORMAL><<EOF>> { BEGIN(INITIAL); }
+
+\"              { BEGIN(STRING); driver.stringLiteral = ""; }
+<STRING>\\\"    { driver.stringLiteral += yytext; }
+<STRING>\\\\    { driver.stringLiteral += yytext; }
+<STRING>\"      { BEGIN(INITIAL);
+                  auto string = cstring(driver.stringLiteral);
+                  return Parser::make_STRING_LITERAL(string, driver.yylloc); }
+<STRING>.       { driver.stringLiteral += yytext; }
+<STRING>\n      { driver.stringLiteral += yytext; }
+
+"abstract"      { BEGIN(NORMAL); return Parser::make_ABSTRACT(driver.yylloc); }
+"action"        { BEGIN(NORMAL); return Parser::make_ACTION(driver.yylloc); }
+"actions"       { BEGIN(NORMAL); return Parser::make_ACTIONS(driver.yylloc); }
+"apply"         { BEGIN(NORMAL); return Parser::make_APPLY(driver.yylloc); }
```

Comment:
One of the changes I've wanted to do was to unify the P4_14 and P4_16 lexers -- but this requires getting rid of the explicit keyword patterns and instead putting all the keywords in a table with their corresponding bison tokens.  The ID rule would then use that table to lookup keywords prior to the symbolkind lookup it already does, and we would just have two static tables (one for P4_14 and one for P4_16).

Unfortunately, it looks like this change would make that very difficult, as the tokens are no longer just enums.